File: Tree/BinarySearchTree/98-validate-binary-search-tree.cpp
/******************************************
 * author      : utoppia
 * description : solution for leetcode
 * updated at  : 2024-01-12 20:09:02
 * filename    : 98-validate-binary-search-tree.cpp
 * language    : cpp
 * editor      : Emacs 27.1
 *****************************************/
#include <algorithm>
#include <bits/stdc++.h>
#include <ios>
#include <iostream>
#include <set>
#include <utility>
#include <vector>

#define vall(v) v.begin(), v.end()
#define pir pair<int, int>

#define pb push_back
#define eb emplace_back
#define fi first
#define snd second
#define sz size

#define ll long long
#define nMax 200005

#define MOD 998244353
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
private:
  bool solve(TreeNode *root, int &maxLeft, int &minRight) {
    if (root->left == NULL and root->right == NULL) {
      maxLeft = minRight = root->val;
      return true;
    }
    int l, r;
    maxLeft = minRight = root->val;
    if (root->left) {
      if (!solve(root->left, l, r))
        return false;
      if (l >= root->val)
        return false;
      maxLeft = max(maxLeft, l);
      minRight = min(minRight, r);
    }
    if (root->right) {
      if (!solve(root->right, l, r))
        return false;
      if (r <= root->val)
        return false;
      maxLeft = max(maxLeft, l);
      minRight = min(minRight, r);
    }
    return true;
  }

public:
  bool isValidBST(TreeNode *root) {
    if (root == NULL)
      return true;
    int l, r;
    return solve(root, l, r);
  }
};

File: Tree/BinarySearchTree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  def dfs(self, root: TreeNode) -> (bool, int, int):
    if not root.left and not root.right:
      return True, root.val, root.val
    x_min, x_max = root.val, root.val
    if root.left:
      is_left, left_min, left_max = self.dfs(root.left)
      if not is_left or left_max >= root.val:
        return False, -1, -1
      x_min = min(x_min, left_min)
      x_max = max(x_max, left_max)
    if root.right:
      is_right, right_min, right_max = self.dfs(root.right)
      if not is_right or right_min <= root.val:
        return False, -1, -1
      x_min = min(x_min, right_min)
      x_max = max(x_max, right_max)
    return True, x_min, x_max

  def isValidBST(self, root: TreeNode) -> bool:
    if not root:
      return True
    ans,_,_ = self.dfs(root)
    return ans
